pfam_experiment.py

```diff
@@ -87,142 +127,171 @@ def main(_):
+
+	embedding_model = create_model(use_transformer=FLAGS.use_transformer, use_bert=FLAGS.use_bert, 
+								   restore_transformer_dir=FLAGS.restore_transformer_dir, encoder_fn=encoder_fn, 
+								   encoder_fn_kwargs=encoder_fn_kwargs, reduce_fn=reduce_fn, reduce_fn_kwargs=reduce_fn_kwargs,
+								   layers=layers, output='embedding')
+
+	embedding_optimizer = create_optimizer(embedding_model,learning_rate=[FLAGS.encoder_lr, FLAGS.lens_lr, FLAGS.predictor_lr], 
+										   weight_decay=[FLAGS.encoder_wd, FLAGS.lens_wd, FLAGS.predictor_wd], layers=layers)
+
+	train_knn_results_untrained_lens = pfam_nearest_neighbors_classification(encoder=embedding_optimizer.target, 
+					                                                             train_family_accessions=lens_knn_train_family_accessions, 
+					                                                             test_family_accessions=lens_knn_train_family_accessions,
+					                                                             batch_size=FLAGS.batch_size,
+					                                                             train_samples=1,
+					                                                             seed=1,
+					                                                             random_state=1)[0]
+	train_knn_accuracy_untrained_lens = train_knn_results_untrained_lens['1-nn accuracy']
+	datum['train_knn_accuracy_untrained_lens_1_knn_train_samples'] = train_knn_accuracy_untrained_lens
+
+	for knn_train_samples in knn_train_samples_:
+
+		test_knn_results_untrained_lens = pfam_nearest_neighbors_classification(encoder=embedding_optimizer.target, 
+					                                                            train_family_accessions=knn_test_family_accessions, 
+					                                                            test_family_accessions=knn_test_family_accessions,
+					                                                            batch_size=FLAGS.batch_size,
+					                                                            train_samples=knn_train_samples,
+					                                                            seed=1,
+					                                                            random_state=1)[0]
+		test_knn_accuracy_untrained_lens = test_knn_results_untrained_lens['1-nn accuracy']
+		datum['test_knn_accuracy_untrained_lens_' + str(knn_train_samples) + '_knn_train_samples'] = test_knn_accuracy_untrained_lens
+
+	encoder_fn_params = None
+	reduce_fn_params = None
+	predict_fn_params = None
+	for i in range(FLAGS.measurements):
+
+		train_batches, train_indexes = create_pfam_batches(family_accessions=lens_knn_train_family_accessions,
+														   batch_size=FLAGS.batch_size,
+														   samples=FLAGS.lens_train_samples,
+														   epochs=measurement_epochs, 
+														   drop_remainder=True)
+
+		model = create_model(use_transformer=FLAGS.use_transformer, use_bert=FLAGS.use_bert, restore_transformer_dir=FLAGS.restore_transformer_dir,
```

Comment:
I'm surprised that you're making new models and optimizers inside the for loop. Why can't you make the model/optimizer just once?

Reply:
when it calls create_model each time, is it making a fresh model with untrained lens parameters? If so, you're essentially restarting training at each iteration of this loop?

Reply:
it's creating the model and optimizer again but using the trained parameters from the previous round of training, I think using optimizer.target instead of model would fix this without having to re-instantiate a 'model' but I didn't want to risk it

Reply:
How does it load in the previous parameters? Is it somehow saved to `restore_transformer_dir`?

Reply:
One concern about reinstantiating (which I think dbelanger mentioned somewhere else) is that you are probably losing the optimizer metadata like moving average of a weight, which could have an effect on optimization.

Reply:
fixed this and will update